nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py

```diff
@@ -829,193 +878,264 @@ def prompt_tuning_param_freeze_and_optimizer_setup(self):
-
-            # bucket prompt tags identically to their corresponding examples
-            if use_soft_prompts:
-                bucket_prompt_tags.append([prompt_tags[item[1]] for item in bucket])
-
-        # Flatten position list
-        positions = [item for sublist in positions for item in sublist]
-
-        # Form request
-        request = {"tokens": buckets, "prompt_tags": bucket_prompt_tags}
-
-        return request, positions, tokens_to_generate, compute_logprobs[0]
-
     def get_parameters(self):
         params = []
         for param_group in self._optimizer_param_groups:
             for param in param_group['params']:
                 params.append(param)
         return params
 
-    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None) -> Any:
-        request, positions, tokens_to_generate, compute_logprobs = MegatronGPTModel._bucketize_gpt_inference(
-            batch, self.use_soft_prompts
-        )
+    def _get_computeprob_response(self, response, inputs):
+        compute_prob_response = {}
+        new_token_ids = []
+        new_tokens = []
+        new_texts = []
+        log_probs = []
+        full_logprobs = []
+        offsets = []
+        for batch_id in range(len(response['tokens'])):
+            if isinstance(inputs, (list, tuple)):
+                if isinstance(inputs[0], str):
+                    new_token_id = self.tokenizer.text_to_ids(inputs[batch_id])
+                    new_text = inputs[batch_id]
+                    token_len = len(new_token_id)
+                elif isinstance(inputs[0], torch.Tensor):
+                    token_len = int(inputs[1][batch_id].item())
+                    new_token_id = inputs[0][batch_id][:token_len].tolist()
+                    new_text = self.tokenizer.ids_to_text(new_token_id)
+            new_token_ids.append(new_token_id)
+            new_tokens.append(response['tokens'][batch_id][:token_len])
+            new_texts.append(new_text)
+            log_probs.append(response['logprob'][batch_id][: (token_len - 1)])
+            full_logprobs.append(response['full_logprob'][batch_id][: (token_len - 1)])
+            offsets.append(response['offsets'][batch_id][:-1])
+        compute_prob_response['sentences'] = new_texts
+        compute_prob_response['tokens'] = new_tokens
+        compute_prob_response['token_ids'] = new_token_ids
+        compute_prob_response['logprob'] = log_probs
+        compute_prob_response['full_logprob'] = full_logprobs
+        compute_prob_response['offsets'] = offsets
+        return compute_prob_response
+
+    def generate(
+        self,
+        inputs: Union[List[str], torch.Tensor, List[dict]],
+        length_params: LengthParam,
+        sampling_params: SamplingParam = None,
+    ) -> OutputType:
+
+        # check whether the DDP is initialized
+        if parallel_state.is_unitialized():
+
+            class RequestDataSet(Dataset):
```

Comment:
Use here the shared class from `dataset_utils`?

Reply:
this is just a hack to invoke the ddp initialization. that's why I use it as an inner class. Need to fix this in the future when we can manually do ddp initialization.